femtools/Particles.F90

```diff
@@ -0,0 +1,1229 @@
+    !!!get name of particle array here to construct the output file
+    particles_cp_filename = trim(prefix)
+    if(present(cp_no)) particles_cp_filename = trim(particles_cp_filename) // "_" // int2str(cp_no)
+    particles_cp_filename = trim(particles_cp_filename) // "_" // trim(lpostfix)
+
+!!!!! Writing of position particles before checkpointing in serial  !!!!!
+
+    !!< Writes particle last position into particles file using MPI output 
+    ! commands so that when running in parallel all processors can write at the same time information into the file at the right location.
+    
+    call MPI_FILE_OPEN(MPI_COMM_FEMTOOLS, trim(particles_cp_filename) // '_particles.' // trim(name) // '.attributes.dat', MPI_MODE_CREATE + MPI_MODE_RDWR, MPI_INFO_NULL, fhdet, IERROR)
+
+    ewrite(1,*) "after opening the IERROR is:", IERROR
+
+    allocate( status(MPI_STATUS_SIZE) )
+
+    call MPI_TYPE_EXTENT(getpreal(), realsize, ierror)
+
+    vfield => extract_vector_field(state(1),"Velocity")
+
+    dimen=vfield%dim
+
+    number_total_columns=num_particles*(dimen+sum(attribute_size))
+
+    node => particle_list%first
+
+    location_to_write=0
+
+    positionloop_cp: do j=1, particle_list%length
+      offset = location_to_write+(node%id_number-1)*(size(node%position)+sum(attribute_size))*realsize
+      ewrite(1,*) "after file set view position IERROR is:", IERROR
+
+      allocate(buffer(size(node%position)+sum(attribute_size)))
+      buffer(1:size(node%position))=node%position
+      if (attribute_size(1).NE.0) then
+         buffer(1+size(node%position):size(node%position)+attribute_size(1))=node%attributes
+      end if
+      if (attribute_size(2).NE.0) then
+         buffer(1+size(node%position)+attribute_size(1):size(node%position)+attribute_size(1) &
+              +attribute_size(2))=node%old_attributes
+      end if
+      if (attribute_size(3).NE.0) then
+         buffer(1+size(node%position)+attribute_size(1)+attribute_size(2):size(node%position)+attribute_size(1) &
+              +attribute_size(2)+attribute_size(3))=node%old_fields
+      end if
+      nints=size(node%position)+sum(attribute_size)
+      
+      call MPI_FILE_WRITE_AT(fhdet,offset,buffer,nints,getpreal(),status,IERROR)
+
+      ewrite(1,*) "after sync position IERROR is:", IERROR
+      deallocate(buffer)
+      node => node%next
+    end do positionloop_cp
+
+    call update_particle_options(trim(particles_cp_filename) // "_particles", "binary", particle_list, name, attribute_size(1), subgroup_path_name)
+    
+    if (fhdet/=0) then
+       call MPI_FILE_CLOSE(fhdet, IERROR)
+       if (IERROR/=0) then
+          ewrite(0,*) "Warning: failed to close .particles checkpoint file open with mpi_file_open"
+       end if
+    end if
+    
+  end subroutine checkpoint_particles
+
+  subroutine update_particle_options(filename, format, particle_list, name, attribute_dims, subgroup_path_name)
+    !!< Updates the initial options of the detectors (options tree in diamond)
```

Comment:
This requires some more explanation: why and what options are we changing?

Hm, this requires some more discussion...